### PR TITLE
Composition: Ensure args get sent to the right ref

### PR DESCRIPTION
**Pull request by tmeasday**
Issue: #10115

## What I did

Made it so that args get sent to and from refs

## How to test

- Is this testable with Jest or Chromatic screenshots?
yes

You can also try adding this to the main.js of official-sb:

```
  refs: {
    'cra-ts-essentials': 'http://localhost:9009',
  },
```

(and running that SB too). Then you can try using args stories in both SBs.

### CODE REVIEW

File: lib/api/src/modules/stories.ts [hunk 1]
```diff
@@ -368,10 +375,36 @@ export const init: ModuleFn = ({
...
+          (storiesHash[storyId] as Story).args = args;
+          store.setState({ storiesHash });
+          break;
+        }
+        case 'external': {
+          const refs = fullAPI.getRefs();
+          const ref = fullAPI.findRef(sourceLocation);
+          (ref.stories[storyId] as Story).args = args;
+          store.setState({ refs: { ...refs, [ref.id]: ref } });
+          break;
```
Comment (tmeasday) [inline lib/api/src/modules/stories.ts:401]: @ndelangen it feels a bit weird to be changing `refs` inside `modules/stories`. But then moving this logic to refs would be spreading the exact same logic (updating args) into two places 🤔 

I think we need to make refs/stories the same thing already!

Comment (ndelangen) [reply lib/api/src/modules/stories.ts:401]: You want to do that before the RC?

I absolutely agree that's where we want to be at some point. I was just being careful of when we do that.

File: lib/api/src/index.tsx [hunk 2]
```diff
@@ -415,7 +415,14 @@ export function useArgs(): [Args, (newArgs: Args) => void] {
   const data = getCurrentStoryData();
   const args = isStory(data) ? data.args : {};
 
-  return [args, (newArgs: Args) => updateStoryArgs(data.id, newArgs)];
+  console.log('in use args');
```
Comment (ndelangen) [inline lib/api/src/index.tsx:418]: remove

File: lib/api/src/modules/stories.ts [hunk 3]
```diff
@@ -368,10 +375,36 @@ export const init: ModuleFn = ({
...
+        }
+        default: {
+          logger.warn('received a STORY_ARGS_UPDATED frame that was not configured as a ref');
+          break;
+        }
```
Comment (ndelangen) [inline lib/api/src/modules/stories.ts:406]: I updated the code here:
https://github.com/storybookjs/storybook/pull/11392

So there's much less boilerplate and duplicated code

File: lib/api/src/modules/stories.ts (hunk 1 above)
Comment (tmeasday) [reply lib/api/src/modules/stories.ts:401]: No I don't really want to do it now. Just wondering what to do about the above code.

File: lib/api/src/modules/stories.ts (hunk 3 above)
Comment (tmeasday) [reply lib/api/src/modules/stories.ts:406]: Yeah I know we can update this afterwards.

File: lib/api/src/modules/stories.ts (hunk 1 above)
Comment (ndelangen) [reply lib/api/src/modules/stories.ts:401]: You could consider using fullAPI.setRef
https://github.com/storybookjs/storybook/blob/9a186bae80359c244a45ed94df47dcb566a51707/lib/api/src/modules/refs.ts#L26-L27
?

Comment (tmeasday) [reply lib/api/src/modules/stories.ts:401]: @ndelangen the implementation of that function expects `stories` to be passed in directly from `SET_STORIES`. Can I refactor it?

Comment (tmeasday) [reply lib/api/src/modules/stories.ts]: @ndelangen https://github.com/storybookjs/storybook/pull/11401/commits/38e66ad7cb59885eb1b6da7fd47aae7c3376a314 -- is that good for you?

File: lib/api/src/modules/refs.ts [hunk 4]
```diff
@@ -213,12 +223,15 @@ export const init: ModuleFn = ({ store, provider, fullAPI }, { runCheck = true }
...
 
+    updateRef: (id, data) => {
+      const { [id]: ref, ...refs } = api.getRefs();
       store.setState({
         refs: {
-          ...api.getRefs(),
-          [id]: result,
+          ...refs,
+          [id]: { ...ref, ...data },
         },
```
Comment (ndelangen) [inline lib/api/src/modules/refs.ts:235]: fancy

File: lib/api/src/modules/stories.ts [hunk 5]
```diff
@@ -283,8 +283,15 @@ export const init: ModuleFn = ({
...
-      fullAPI.emit(UPDATE_STORY_ARGS, id, newArgs);
+    updateStoryArgs: (story, updatedArgs) => {
+      const { id: storyId, refId } = story;
+      fullAPI.emit(UPDATE_STORY_ARGS, {
+        storyId,
+        updatedArgs,
+        options: {
+          target: refId ? `storybook-ref-${refId}` : 'storybook-preview-iframe',
+        },
+      });
```
Comment (ndelangen) [inline lib/api/src/modules/stories.ts:294]: 👍

File: lib/api/src/modules/stories.ts [hunk 6]
```diff
@@ -397,10 +404,35 @@ export const init: ModuleFn = ({
...
+    fullAPI.on(STORY_ARGS_UPDATED, function handleStoryArgsUpdated({
+      storyId,
+      args,
+    }: {
+      storyId: StoryId;
+      args: Args;
+    }) {
+      // the event originates from an iframe, event.source is the iframe's location origin + pathname
+      const { source }: { source: string } = this;
+      const [sourceType, sourceLocation] = getSourceType(source);
```
Comment (ndelangen) [inline lib/api/src/modules/stories.ts:416]: this will be cleaned up a bit in my PR 👍

Comment (tmeasday) [reply lib/api/src/modules/stories.ts:416]: It's a race to see who gets merged first

**Review by ndelangen — APPROVED** [review]

**Review by shilman — APPROVED** [review]
Looks great

**Pull request merged**